### ISSUE TITLE
Playerdata API

### ISSUE THREAD

**Issue by michelevantaggi02** [R2Northstar/NorthstarMasterServer#31]
Get persistent player data from the database in JSON by HTTPS request.
It could be useful to build dashboards and leaderboards.

**Comment by michelevantaggi02** [comment]
If it's possible to do i can also help, but i still have to start with nodejs

**Comment by michelevantaggi02** [comment]
@barnabwhy That's neat, i started looking at the json form of the playerdata to look for what i could get and use for the API, but without any game played on my master server i only have default values and only my account.
It could be useful to have a minimal part (like 5 to 10 accounts) of the real master server, but i have seen that there are ip addresses and i think origin tokens (didn't figure out what the id col in the database is really), so i understand i could be kind of a trouble to use

**Comment by michelevantaggi02** [comment]
@barnabwhy i also have seen on your branch that you too have trouble with usernames, so i'm not the only one who found it strange there isn't really a proper attribute on the player table

**Comment by barnabwhy** [comment]
> @barnabwhy i also have seen on your branch that you too had trouble with usernames (i've seen the pull request), so i'm not the only one who found it strange there isn't really a proper attribute on the player table

I made the PR that adds username stuff, I added a comment in the player api stuff just so i don't forget i can add it once that pr is merged

**Comment by barnabwhy** [comment]
> @barnabwhy That's neat, i started looking at the json form of the playerdata to look for what i could get and use for the API, but without any game played on my master server i only have default values and only my account. It could be useful to have a minimal part (like 5 to 10 accounts) of the real master server, but i have seen that there are ip addresses and i think origin tokens (didn't figure out what the id col in the database is really), so i understand it could be kind of a trouble to use

id is just a uuid generated by the master server on first auth and origin tokens aren't stored. also, yes, ip addresses are stored in the table but that isn't just gonna be put on an endpoint. i'm ideally gonna make stuff that has cherry-picked properties

**Comment by michelevantaggi02** [comment]
> > @barnabwhy That's neat, i started looking at the json form of the playerdata to look for what i could get and use for the API, but without any game played on my master server i only have default values and only my account. It could be useful to have a minimal part (like 5 to 10 accounts) of the real master server, but i have seen that there are ip addresses and i think origin tokens (didn't figure out what the id col in the database is really), so i understand it could be kind of a trouble to use
> 
> id is just a uuid set by the master server on first auth (origin player id) and origin tokens aren't stored. also, yes, ip addresses are stored in the table but that isn't just gonna be put on an endpoint. i'm ideally gonna make stuff that has cherry-picked properties

Yeah that's what i was thinking, there is no reason for an api to exists if you can make any sort of request, i made a little list of things you could ask to the API maybe i can write it here so we can work on them, they are mainly items and player leaderboards and single accounts stuff

**Comment by michelevantaggi02** [comment]
Account info:

- [ ] Username
- [ ] Faction
- [ ] Last game(s) played
- [ ] Current loadout
- [ ] Current titan
- [ ] K/D
- [ ] Most used weapon
- [ ] Most kill in single match
- [ ] Total kills
- [ ] Total deaths
- [ ] Stats per weapon (?)

**Comment by barnabwhy** [comment]
I'd say it's probably better to have some of those split off into loadout and stats endpoints

**Comment by michelevantaggi02** [comment]
I'm not sure about the leaderboard stuff, because you have to check every account, maybe i could write a script that updates a separate table every time a player ends a match, but i understand it could be a lot of resource usage

**Comment by barnabwhy** [comment]
maybe a routine check or something

**Comment by michelevantaggi02** [comment]
> I'd say it's probably better to have some of those split off into loadout and stats endpoints

Yeah, it won't be much of a trouble to split it up or get everything depending on the GET request params, i've already done something similar in the past (not in JSON but with SQL queries)

**Comment by michelevantaggi02** [comment]
> maybe a routine check or something

I'm kinda new to nodejs, didn't now it is a thing

**Comment by barnabwhy** [comment]
The things that [my branch](https://github.com/barnabwhy/NorthstarMasterServer/tree/pdata-json-api) returns through GET requests:

**/player/pdata**
* [x] All pdata (excluding pdiff)
- Pretty much encompasses all the below and more, the other ones are for getting more specific information
- This endpoint notably **excludes** username

**/player/info**
* [x]  Username (if #42 gets merged)
* [x]  Player id
* [x]  Gen
* [x]  XP
* [x] Calling card
* [x] Callsign icon
* [x] Net worth


**/player/stats**
* [x]  Player id
* [x]  Gen
* [x]  XP
* [x] Credits
* [x] Net worth
* [x] Faction XP
* [x] Titan XP
* [x] FD titan XP
* [x] Game stats
* [x] Map stats
* [x] Time stats
* [x] Distance stats
* [x] Weapon stats
* [x] Weapon kill stats
* [x] Kill stats
* [x] Death stats
* [x] Misc stats
* [x] FD stats
* [x] Titan stats
* [x] KD ratio
* [x] PVP KD ratio
* [x] Win streak
* [x] Highest win streak


**/player/loadout**
* [x]  Faction
* [x] Full active pilot loadout
* [x] Full active titan loudout
* [x] Full pilot loadout list
* [x] Full titan loadout list

**Comment by abarichello** [comment]
Closed by #48
